### PR TITLE
Feature (hashes): Add from_bytes_ref and from_bytes_mut to all Hash types

### LINKED ISSUE #1756: Guarantee memory layout of hashes and support casts?
Being able to cast `&[u8; N]` to an `N`-byte hash type could be very helpful for fast parsing like [`bitcoin_slices`](https://github.com/RCasatta/bitcoin_slices) is doing. To enable this the hash types must be `#[repr(transparent)]` which I think we could support since I don't see any good reason  why the memory layout would ever change.

Further, we could provide these conversions:

```rust
fn ref_from_bytes(bytes: &[u8; $len]) -> &Self {
    // Sound because Self is #[repr(transparent)] containing [u8; $len]
    unsafe { &*(bytes as *const _ as *const Self) }
}

fn mut_from_bytes(bytes: &mut [u8; $len]) -> &mut Self {
    // Sound because Self is #[repr(transparent)] containing [u8; $len]
    unsafe { &*(bytes as *mut _ as *mut Self) }
}
```

`unsafe` here is even more trivial than that in `Script`, so should be OK.

### PR DESCRIPTION
**Pull request by junderw**
Closes #1756

I added tests to one of the Hash types, let me know if this seems reasonable as a test and I'll add it for all the other Hash types.

### CODE REVIEW

**Review by apoelstra — APPROVED** [review]
ACK 91863c8a785e4e0faf7cfc6f6a5f607cf48b4833

**Comment by junderw** [comment]
Any thoughts on the tests? (Do we need tests for each Hash type or is just one enough?)

**Comment by tcharding** [comment]
> Any thoughts on the tests? (Do we need tests for each Hash type or is just one enough?)

IMO the tests are due for a refactor since they have loads of duplicate code. You can leave this PR as is though if you want and I'll add cleaning them up to my TODO list. I'll wait for your reply before adding it to my list.

FTR we don't tend to use "Feature" pre-fixes in the commit brief message, just `hashes: ...` will do. No one is super particular about it though so feel free to do as you please. Thanks for the work!

**Review by tcharding — APPROVED** [review]
ACK 91863c8a785e4e0faf7cfc6f6a5f607cf48b4833

**Comment by apoelstra** [comment]
Will wait for @Kixunil before merging, since it's his issue -- and also I'm uncertain if the `from_bytes_ref` name is what we want according to the rust guidelines.

**Comment by junderw** [comment]
> I'll wait for your reply before adding it to my list.

Sure thing. Go ahead.

File: hashes/src/internal_macros.rs
Comment (Kixunil) [inline hashes/src/internal_macros.rs:100]: I think this is technically unsound because someone could call the macro outside of `hash_newtype` and get this functions implemented. Maybe I'm paranoid but I'd rather have it in `hash_type` macro.
And while you're at it add a non-doc comment: `#[repr(transaparent)] // required for soundness of from_bytes_ref/mut`, so that nobody removes it.

Also it'd be nice to document what `#[repr(transparent)]` means for consumers: The memory layout is guaranteed to be the same as `[u8; LEN]`.

File: hashes/src/ripemd160.rs
Comment (Kixunil) [inline hashes/src/ripemd160.rs:487]: Wouldn't it be better to just change `Vec<u8>` to `[u8; 20]` above to avoid this noise of converting?

Also, ideally these tests should run through `miri` but we might need to `cfg(not(miri))` other some things.

**Review by Kixunil — COMMENTED** [review]
Oh wow, I totally missed it already was `#[repr(transparent)]` although technically unusable because it wasn't documented.

**Comment by Kixunil** [comment]
To clarify, the PR is not technically wrong, just a bit foot-gunny. We can improve it in a later PR if you want.

File: hashes/src/internal_macros.rs
Comment (apoelstra) [reply hashes/src/internal_macros.rs:100]: Ooh, good catch! I dutifully checked that every caller of this macro had `#[repr(transparent)]` (and was also surprised, because this isn't documented and it's actually hard to find the exact definitions of everything when everything is called `Hash` and defined in a macro soup). But of course, we are not necessarily the only callers of the macro.

**Comment by apoelstra** [comment]
I'm gonna go ahead and merge it. IIRC you have another issue/PR about `repr(transparent)` which we can repurpose to move these functions to a safer place.

**Comment by Kixunil** [comment]
Sure.

**Comment by tcharding** [comment]
> also I'm uncertain if the from_bytes_ref name is what we want according to the rust guidelines.

No guarantee that I got it right but FWIW I spent 5 minutes going through docs looking to see we had the naming correct.

**Pull request merged**
